File: apps/worker/scripts/migrate.py
import os
import glob
import psycopg
from dotenv import load_dotenv

ENV_PATH = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".env"))
load_dotenv(dotenv_path=ENV_PATH)

DATABASE_URL = os.getenv("DATABASE_URL", "").strip()
if not DATABASE_URL:
    raise RuntimeError(f"DATABASE_URL is required (check {ENV_PATH})")

MIGRATIONS_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "migrations"))

def ensure_table(conn):
    conn.execute("""
    CREATE TABLE IF NOT EXISTS schema_migrations (
      version TEXT PRIMARY KEY,
      applied_at TIMESTAMPTZ NOT NULL DEFAULT now()
    );
    """)

def applied_versions(conn) -> set[str]:
    rows = conn.execute("SELECT version FROM schema_migrations").fetchall()
    return {r[0] for r in rows}

def main():
    files = sorted(glob.glob(os.path.join(MIGRATIONS_DIR, "*.sql")))
    if not files:
        print("No migrations found.")
        return

    with psycopg.connect(DATABASE_URL) as conn:
        conn.autocommit = False
        ensure_table(conn)
        done = applied_versions(conn)

        for path in files:
            version = os.path.basename(path)
            if version in done:
                continue

            sql = open(path, "r", encoding="utf-8").read()
            print(f"Applying {version}...")
            conn.execute(sql)
            conn.execute("INSERT INTO schema_migrations (version) VALUES (%s)", (version,))
            conn.commit()

    print("Migrations complete.")

if __name__ == "__main__":
    main()

File: packages/db/migrate.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { spawnSync } from "node:child_process";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// repo root (where docker-compose.yml lives)
const repoRoot = path.resolve(__dirname, "..", "..");
const sqlDir = path.resolve(__dirname, "sql");

function runDockerPsql(args, input = "") {
  const base = [
    "compose",
    "exec",
    "-T",
    "db",
    "psql",
    "-U",
    "postgres",
    "-d",
    "insider_radar",
    "-v",
    "ON_ERROR_STOP=1",
    ...args,
  ];

  const r = spawnSync("docker", base, {
    cwd: repoRoot,
    encoding: "utf8",
    input,
  });

  if (r.error) throw r.error;
  if (r.status !== 0) {
    const msg = (r.stderr || r.stdout || "").trim();
    throw new Error(`psql failed (exit ${r.status}): ${msg}`);
  }

  return (r.stdout || "").trim();
}

function psqlQuery(sql) {
  // -tA = tuples only, unaligned (easy to parse)
  return runDockerPsql(["-tA", "-c", sql]);
}

async function main() {
  // ensure migrations table exists
  runDockerPsql([
    "-c",
    `
    create table if not exists _migrations (
      id bigserial primary key,
      filename text not null unique,
      applied_at timestamptz not null default now()
    );
    `,
  ]);

  const files = fs
    .readdirSync(sqlDir)
    .filter((f) => f.endsWith(".sql"))
    .sort();

  for (const f of files) {
    const safeName = f.replaceAll("'", "''");
    const already = psqlQuery(
      `select 1 from _migrations where filename='${safeName}' limit 1;`
    );

    if (already === "1") {
      console.log("Skipping", f, "(already applied)");
      continue;
    }

    const fullPath = path.join(sqlDir, f);
    let sql = fs.readFileSync(fullPath, "utf8");
    sql = sql.replace(/^\uFEFF/, ""); // strip UTF-8 BOM if present

    console.log("Applying", f);

    // Apply in a transaction and record the migration
    const script = `
begin;
${sql}
insert into _migrations(filename) values ('${safeName}');
commit;
`;

    // Feed script via stdin: psql -f -
    runDockerPsql(["-f", "-"], script);

    console.log("Applied", f);
  }

  console.log("Migrations complete.");
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

